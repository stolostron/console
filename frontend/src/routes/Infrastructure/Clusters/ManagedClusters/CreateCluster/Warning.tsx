/* Copyright Contributors to the Open Cluster Management project */

import { Alert, AlertVariant, ButtonVariant } from '@patternfly/react-core'
import { AcmButton } from '../../../../../ui-components'
import { Link } from 'react-router-dom-v5-compat'
import { createContext, useContext } from 'react'

export type WarningContextType =
  | { title: string; text: string; linkTo?: string; linkText?: string; isExternalLink?: boolean }
  | undefined

export const WarningContext = createContext<WarningContextType>(undefined)

export const Warning = () => {
  const warning = useContext(WarningContext)

  if (!warning) {
    return null
  }

  let actionLinks
  if (warning.linkTo) {
    warning.isExternalLink
      ? (actionLinks = (
          <a href={warning.linkTo} target="_blank" rel="noopener noreferrer" style={{ paddingLeft: 0 }}>
            {warning.linkText}
          </a>
        ))
      : (actionLinks = (
          <Link to={warning.linkTo}>
            <AcmButton variant={ButtonVariant.link} style={{ paddingLeft: 0 }}>
              {warning.linkText}
            </AcmButton>
          </Link>
        ))
  }

  return (
    <>
      <Alert title={warning?.title} variant={AlertVariant.warning} actionLinks={actionLinks} isInline>
        {warning.text}
      </Alert>
      <br />
    </>
  )
}
