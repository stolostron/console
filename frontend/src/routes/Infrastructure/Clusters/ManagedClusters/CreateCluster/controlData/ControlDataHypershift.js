/* Copyright Contributors to the Open Cluster Management project */
import React from 'react'
import DetailsForm from '../components/assisted-installer/hypershift/DetailsForm'
import HostsForm from '../components/assisted-installer/hypershift/HostsForm'
import NetworkForm from '../components/assisted-installer/hypershift/NetworkForm'
import {
    automationControlData,
    appendKlusterletAddonConfig,
    appendWarning,
    insertToggleModalFunction,
} from './ControlDataHelpers'
import { CreateCredentialModal } from '../../../../../../components/CreateCredentialModal'

export const getControlDataHypershift = (
    handleModalToggle,
    warning,
    includeAutomation = true,
    includeKlusterletAddonConfig = true
) => {
    const controlData = [...controlDataHypershift]
    appendKlusterletAddonConfig(includeKlusterletAddonConfig, controlData)
    insertToggleModalFunction(handleModalToggle, controlData)
    if (warning) {
        appendWarning(warning, controlData)
    }
    if (includeAutomation) {
        return [...controlData, ...automationControlData]
    }
    return controlData
}

const controlDataHypershift = [
    ////////////////////////////////////////////////////////////////////////////////////
    ///////////////////////  AI form  /////////////////////////////////////
    {
        id: 'hypershiftDetailStep',
        type: 'step',
        title: 'Cluster details',
    },
    {
        id: 'infrastructure',
        name: 'Infrastructure',
        active: 'Host inventory',
        type: 'reviewinfo',
    },
    {
        id: 'controlplane',
        name: 'Control plane type',
        active: 'Hosted',
        type: 'reviewinfo',
    },
    /////////////////////// ACM Credentials  /////////////////////////////////////
    {
        name: 'creation.ocp.cloud.connection',
        tooltip: 'tooltip.creation.ocp.cloud.connection',
        id: 'connection',
        type: 'singleselect',
        placeholder: 'creation.ocp.cloud.select.connection',
        providerId: 'hostinventory',
        validation: {
            notification: 'creation.ocp.cluster.must.select.connection',
            required: false,
        },
        available: [],
        footer: <CreateCredentialModal />,
    },
    {
        id: 'hypershift',
        type: 'custom',
        component: <DetailsForm />,
        providerId: 'hypershift',
        mustValidate: true,
        encodeValues: ['pullSecret'],
        additionalProps: {
            promptSshPublicKey: false,
        },
    },
    {
        id: 'hypershiftHostsStep',
        type: 'step',
        title: 'Node pools',
        disabled: true,
    },
    {
        id: 'hypershift-hosts',
        type: 'custom',
        component: <HostsForm />,
        providerId: 'hypershift',
        mustValidate: true,
    },
    {
        id: 'hyperhisftNetworkStep',
        type: 'step',
        title: 'Networking',
        disabled: true,
    },
    {
        id: 'hypershift-network',
        type: 'custom',
        component: <NetworkForm />,
        providerId: 'hypershift',
        mustValidate: true,
    },
]

export default getControlDataHypershift
