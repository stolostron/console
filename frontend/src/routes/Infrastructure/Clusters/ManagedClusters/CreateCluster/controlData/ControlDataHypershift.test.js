// Copyright (c) 2022 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
'use strict'

import { Warning } from '../Warning'
import { getControlDataHypershift } from './ControlDataHypershift'
import i18next from 'i18next'

const t = i18next.t.bind(i18next)
const handleModalToggle = jest.fn()

describe('Cluster creation control data for Hypershift', () => {
  it('generates correctly', () => {
    expect(getControlDataHypershift(t, handleModalToggle, <Warning />, true, true)).toMatchSnapshot()
  })
  it('generates correctly for MCE', () => {
    expect(getControlDataHypershift(t, handleModalToggle, <Warning />, true, false)).toMatchSnapshot()
  })
})
