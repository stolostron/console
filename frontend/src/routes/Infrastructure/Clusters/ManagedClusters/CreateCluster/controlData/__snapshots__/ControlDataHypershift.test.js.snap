// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cluster creation control data for Hypershift generates correctly 1`] = `
Array [
  Object {
    "id": "hypershiftDetailStep",
    "title": "Cluster details",
    "type": "step",
  },
  Object {
    "active": "Host inventory",
    "id": "infrastructure",
    "name": "Infrastructure",
    "type": "reviewinfo",
  },
  Object {
    "active": "Hosted",
    "id": "controlplane",
    "name": "Control plane type",
    "type": "reviewinfo",
  },
  Object {
    "available": Array [],
    "footer": <CreateCredentialModal
      handleModalToggle={[MockFunction]}
    />,
    "id": "connection",
    "name": "Infrastructure provider credential",
    "placeholder": "Select a credential",
    "providerId": "hostinventory",
    "tooltip": "The settings that are required for the selected provider. You can select an existing connection, or add a new connection. Cannot be changed after creation.",
    "type": "singleselect",
    "validation": Object {
      "notification": "Select a connection",
      "required": false,
    },
  },
  Object {
    "additionalProps": Object {
      "promptSshPublicKey": false,
    },
    "component": <DetailsForm />,
    "encodeValues": Array [
      "pullSecret",
    ],
    "id": "hypershift",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "disabled": true,
    "id": "hypershiftHostsStep",
    "title": "Node pools",
    "type": "step",
  },
  Object {
    "component": <HostsForm />,
    "id": "hypershift-hosts",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "disabled": true,
    "id": "hyperhisftNetworkStep",
    "title": "Networking",
    "type": "step",
  },
  Object {
    "component": <NetworkForm />,
    "id": "hypershift-network",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "active": true,
    "id": "includeKlusterletAddonConfig",
    "type": "hidden",
  },
  Object {
    "component": <Warning />,
    "id": "warning",
    "type": "custom",
  },
  Object {
    "id": "automationStep",
    "title": "Automation",
    "type": "step",
  },
  Object {
    "id": "chooseTemplate",
    "info": "Optional: Choose an automation template to run Ansible jobs at different stages of the cluster lifecycle.",
    "type": "title",
  },
  Object {
    "component": <AutomationProviderHint
      className="creation-view-controls-hint"
      component="hint"
    />,
    "id": "automationProviderHint",
    "type": "custom",
  },
  Object {
    "id": "templateName",
    "name": "Automation template",
    "onSelect": [Function],
    "placeholder": "Select an automation template",
    "prompts": Object {
      "icon": <ExternalLinkAltIcon
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
      />,
      "id": "add-automation-template",
      "positionBottomRight": true,
      "prompt": "Add automation template",
      "type": "link",
      "url": "/multicloud/infrastructure/automations/add",
    },
    "tooltip": "Only templates with valid Ansible Automation Platform credentials are listed.",
    "type": "singleselect",
    "validation": Object {
      "required": false,
    },
  },
  Object {
    "component": <TemplateLinkOutControl />,
    "id": "curatorLinkOut",
    "type": "custom",
  },
  Object {
    "component": <TemplateSummaryControl />,
    "id": "curatorSummary",
    "type": "custom",
  },
  Object {
    "active": "",
    "id": "clusterCuratorSpec",
    "reverse": [Function],
    "type": "hidden",
  },
  Object {
    "active": Array [],
    "hidden": true,
    "id": "supportedCurations",
    "type": "values",
  },
  Object {
    "active": "",
    "id": "toweraccess-install",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-upgrade",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-scale",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-destroy",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "reconcilePause",
    "type": "hidden",
  },
]
`;

exports[`Cluster creation control data for Hypershift generates correctly for MCE 1`] = `
Array [
  Object {
    "id": "hypershiftDetailStep",
    "title": "Cluster details",
    "type": "step",
  },
  Object {
    "active": "Host inventory",
    "id": "infrastructure",
    "name": "Infrastructure",
    "type": "reviewinfo",
  },
  Object {
    "active": "Hosted",
    "id": "controlplane",
    "name": "Control plane type",
    "type": "reviewinfo",
  },
  Object {
    "available": Array [],
    "footer": <CreateCredentialModal
      handleModalToggle={[MockFunction]}
    />,
    "id": "connection",
    "name": "Infrastructure provider credential",
    "placeholder": "Select a credential",
    "providerId": "hostinventory",
    "tooltip": "The settings that are required for the selected provider. You can select an existing connection, or add a new connection. Cannot be changed after creation.",
    "type": "singleselect",
    "validation": Object {
      "notification": "Select a connection",
      "required": false,
    },
  },
  Object {
    "additionalProps": Object {
      "promptSshPublicKey": false,
    },
    "component": <DetailsForm />,
    "encodeValues": Array [
      "pullSecret",
    ],
    "id": "hypershift",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "disabled": true,
    "id": "hypershiftHostsStep",
    "title": "Node pools",
    "type": "step",
  },
  Object {
    "component": <HostsForm />,
    "id": "hypershift-hosts",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "disabled": true,
    "id": "hyperhisftNetworkStep",
    "title": "Networking",
    "type": "step",
  },
  Object {
    "component": <NetworkForm />,
    "id": "hypershift-network",
    "mustValidate": true,
    "providerId": "hypershift",
    "type": "custom",
  },
  Object {
    "active": false,
    "id": "includeKlusterletAddonConfig",
    "type": "hidden",
  },
  Object {
    "component": <Warning />,
    "id": "warning",
    "type": "custom",
  },
  Object {
    "id": "automationStep",
    "title": "Automation",
    "type": "step",
  },
  Object {
    "id": "chooseTemplate",
    "info": "Optional: Choose an automation template to run Ansible jobs at different stages of the cluster lifecycle.",
    "type": "title",
  },
  Object {
    "component": <AutomationProviderHint
      className="creation-view-controls-hint"
      component="hint"
    />,
    "id": "automationProviderHint",
    "type": "custom",
  },
  Object {
    "id": "templateName",
    "name": "Automation template",
    "onSelect": [Function],
    "placeholder": "Select an automation template",
    "prompts": Object {
      "icon": <ExternalLinkAltIcon
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
      />,
      "id": "add-automation-template",
      "positionBottomRight": true,
      "prompt": "Add automation template",
      "type": "link",
      "url": "/multicloud/infrastructure/automations/add",
    },
    "tooltip": "Only templates with valid Ansible Automation Platform credentials are listed.",
    "type": "singleselect",
    "validation": Object {
      "required": false,
    },
  },
  Object {
    "component": <TemplateLinkOutControl />,
    "id": "curatorLinkOut",
    "type": "custom",
  },
  Object {
    "component": <TemplateSummaryControl />,
    "id": "curatorSummary",
    "type": "custom",
  },
  Object {
    "active": "",
    "id": "clusterCuratorSpec",
    "reverse": [Function],
    "type": "hidden",
  },
  Object {
    "active": Array [],
    "hidden": true,
    "id": "supportedCurations",
    "type": "values",
  },
  Object {
    "active": "",
    "id": "toweraccess-install",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-upgrade",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-scale",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "toweraccess-destroy",
    "type": "hidden",
  },
  Object {
    "active": "",
    "id": "reconcilePause",
    "type": "hidden",
  },
]
`;
